benchmarks/serialize.py

```diff
@@ -0,0 +1,246 @@
+
+    Returns:
+        A callable for the hashing factory.
+    """
+    hash_engine = get_hash_engine_factory(hash_algorithm)
+
+    def _hasher_factory(
+        path: pathlib.Path, start: int, end: int
+    ) -> file.ShardedFileHasher:
+        return file.ShardedFileHasher(
+            path,
+            hash_engine(),  # pytype: disable=not-instantiable
+            start=start,
+            end=end,
+            chunk_size=chunk_size,
+            shard_size=shard_size,
+        )
+
+    return _hasher_factory
+
+
+def get_file_hasher_factory(
+    hash_algorithm: str, chunk_size: int
+) -> Callable[[pathlib.Path], file.FileHasher]:
+    """Returns a hasher factory for file serialization.
+
+    Args:
+        hash_algorithm: the hash algorithm to use for each file.
+        chunk_size: the chunk size to use when reading files.
+
+    Returns:
+        A callable for the hashing factory.
+    """
+    hash_engine = get_hash_engine_factory(hash_algorithm)
+
+    def _hasher_factory(path: pathlib.Path) -> file.FileHasher:
+        return file.SimpleFileHasher(
+            path,
+            hash_engine(),  # pytype: disable=not-instantiable
+            chunk_size=chunk_size,
+        )
+
+    return _hasher_factory
+
+
+def run(args: argparse.Namespace) -> None:
+    """Performs the benchmark.
+
+    Args:
+        args: The arguments specifying the benchmark scenario.
+    """
+    # 1. Hashing layer
+    if args.use_shards:
+        hasher = get_sharded_file_hasher_factory(
+            args.hash_method, args.chunk, args.shard
+        )
+    else:
+        hasher = get_file_hasher_factory(args.hash_method, args.chunk)
+
+    # 2. Serialization layer
+    if args.skip_manifest or args.single_digest:
+        merge_hasher_factory = get_hash_engine_factory(args.merge_hasher)
+        if args.use_shards:
+            serializer = serialize_by_file_shard.DigestSerializer(
+                hasher,
+                merge_hasher_factory(),  # pytype: disable=not-instantiable
+                max_workers=args.max_workers,
+            )
+        else:
+            # This gets complicated because the API here is not matching the
+            # rest. We should fix this.
+            if args.max_workers is not None and args.max_workers != 1:
+                raise ValueError("Currently, only 1 worker is supported here")
+            serializer = serialize_by_file.DigestSerializer(
+                # pytype: disable=wrong-arg-count
+                hasher(pathlib.Path("unused")),
+                # pytype: enable=wrong-arg-count
+                merge_hasher_factory,
+            )
+    else:
+        if args.use_shards:
+            serializer_factory = serialize_by_file_shard.ManifestSerializer
+        else:
+            serializer_factory = serialize_by_file.ManifestSerializer
+
+        serializer = serializer_factory(hasher, max_workers=args.max_workers)
+
+    # 3. Signing layer
+    if args.skip_manifest:
+        in_toto_builder = id
```

Comment:
Nice catch. I wanted `lambda x: x` which is `id` in Haskell. I'll change

Reply:
they're almost identical, I believe the comment was mainly from me having to pause for a second to figure out the scoping, especially since `in_toto_builder` is a function.

Reply:
Yeah, that's why I didn't notice it's not doing what it's supposed to. Desired was a function that returns its argument instead got a function that returns "an address" of the argument. Both are fast and the argument is the expensive part, so no way to notice it